[exclusivity] Add support for Task local exclusivity access sets.

gottesmm opened the pull request:
The implemented semantics are that:

1. Tasks have separate exclusivity access sets.
2. Any synchronous context that creates tasks will have its exclusive access set
   merged into the Tasks while the Task is running.

rdar://80492364

gottesmm committed:
[exclusivity] Add support for Task local exclusivity access sets.

The implemented semantics are that:

1. Tasks have separate exclusivity access sets.
2. Any synchronous context that creates tasks will have its exclusive access set
   merged into the Tasks while the Task is running.

rdar://80492364

gottesmm commented:
Just an initial version of this to get feedback before I wire things up/create tests/etc.

gottesmm commented on stdlib/public/runtime/Exclusivity.cpp:
```diff
@@ -222,6 +228,20 @@ class AccessSet {
   }
 
+  /// Return the parent access of \p childAccess in the list.
+  Access *findParentAccess(Access *childAccess) {
```
I need to take another look at this. I am a little worried that I may not handle the corner case where childAccess is the first case. I just need to think about it after I eat.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
Just to make sure things compile.

gottesmm commented:
@swift-ci smoke test

gottesmm commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -379,8 +388,10 @@ inline void AsyncTask::flagAsSuspended() {
                                                 std::memory_order_relaxed,
-                                                std::memory_order_relaxed))
+                                                std::memory_order_relaxed)) {
+      swift_task_exitThreadLocalContext((char *)&_private().ExclusivityAccessSet[0]);
```
@rjmccall I am pretty sure this is the wrong way to do this. I just put it here for testing purposes. Can you tell me the right place to do this?

gottesmm commented on stdlib/public/runtime/Exclusivity.cpp:
```diff
@@ -202,6 +208,7 @@ class AccessSet {
   }
 
   void remove(Access *access) {
+    assert(Head != nullptr && "Can not remove access from empty access list");
```
I put in this assert so we always get a nice msg when we are in this invalid state.

kavon replied:
FYI I have the same assert planned for https://github.com/apple/swift/pull/38215 which is close to being merged (just need to fix a test on Linux).

gottesmm replied:
Sounds good! I'll be sure to rebase. I think this is going to take longer.

gottesmm commented on test/Concurrency/Runtime/exclusivity.swift:
```diff
@@ -0,0 +1,42 @@
+      print("In my main")
+    }
+
+    static func myMain2(_ x: inout Int, _ shouldSleep: Bool) async {
```
I am going to add some more tests that exercise this where there are multiple elements in the exclusivity set we need to move and also the synchronous case we talked about.

gottesmm committed:
[concurrency] Add AsyncTask::flagAsCompleted() and behind SWIFT_TASK_PRINTF_DEBUG allow it to be used to log Task completion.

This should let people debug Task lifetime easier.

gottesmm committed:
[exclusivity] Introduce an env variable to in asserts builds cause the exclusivity part of the runtime to emit logging information.

The specific environment variable is SWIFT_DEBUG_RUNTIME_EXCLUSIVITY_LOGGING. We only
check the environment the first time we try to lookup the TLSContext, so it
shouldn't impact exclusivity performance.

My intention is to use this to write some FileCheck tests so that we can really
be sure that the runtime is doing what we think it is. As such I also changed
the small amount of logging routines meant for use in the debugger to use stdout.

gottesmm committed:
[exclusivity] Add tests using the new exclusivity logging in asserts builds to more rigorously test the Task based swizzling AccessSet functionality.

I was writing a bunch of functional tests and realized I needed to be more
rigorous and get actual runtime logging to be able to be sure that the runtime
feature was working correctly.

gottesmm commented:
@swift-ci smoke test

mikeash commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -282,14 +283,23 @@ struct AsyncTask::PrivateStorage {
+  /// but actually initialize its bit state to a const global provided by
+  /// libswiftCore so that libswiftCore can control the layout of our initial
+  /// state.
+  uintptr_t ExclusivityAccessSet[2] = {0,0};
```
I think this might exceed the reserved space on 32-bit. You may need to expand it there.

gottesmm replied:
@mikeash I was talking with @rjmccall at one point and he mentioned that I might be able to use different storage rather than in task itself, but I am not sure exactly what he was talking about.

mikeash commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -379,11 +395,16 @@ inline void AsyncTask::flagAsSuspended() {
 }
 
+inline void AsyncTask::flagAsCompleted() {
+}
```
This function intentionally left blank?

gottesmm replied:
Yes. It is meant to be a place where you could put logging. I was nervous about putting it in. But maybe I should?

mikeash replied:
Might make sense to preemptively put in the logging, ifdefed out (`SWIFT_TASK_PRINTF_DEBUG` is already used elsewhere and would probably make sense here). Then it makes the purpose of this function apparent, _and_ helps out whoever needs some logging in the future.

mikeash commented on stdlib/public/runtime/Exclusivity.cpp:
```diff
@@ -171,8 +182,18 @@ class AccessSet {
   bool insert(Access *access, void *pc, void *pointer, ExclusivityFlags flags) {
+#ifndef NDEBUG
+    if (isExclusivityLoggingEnabled())
+      printf("Inserting new access: %p\n", access);
```
Outputting these to stderr would probably be a little better.

gottesmm replied:
You are right. I was doing it to help me writing the tests. But I was being lazy. I'll fix it.

gottesmm commented:
@swift-ci test

gottesmm commented:
Doing the full test to test the 32 bit issue

gottesmm committed:
[concurrency] Add UnownedJob._runSynchronously(on:) to support running a job synchronously on an executor.

I am going to use this to test that we propagate synchronous accesses into
asynchronous tasks access sets.

To ensure this is not ABI, I underscored/marked this as alwaysEmitIntoClient.

gottesmm committed:
[exclusivity] Add logging lock/flushing to prevent output interleaving.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test windows platform

gottesmm commented:
The simulator builds are going to fail since I forgot to define SIMCTL_CHILD_SWIFT_DEBUG_RUNTIME_EXCLUSIVITY_LOGGING along side SWIFT_DEBUG_RUNTIME_EXCLUSIVITY_LOGGING .

gottesmm commented:
But I am going to wait for the Linux/Windows builds to finish since I am curious how far they get.

gottesmm commented:
I think there may be an issue with one of my tests on Windows. I am going to push so I can restart the full test on macOS.

gottesmm commented:
Tests passed on Linux both w/without opts.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test Apple Silicon

gottesmm commented:
Looks like case8 with custom executors hung.

gottesmm commented:
Apple Silicon Test is saying that I didn't mark something with proper availability.

gottesmm committed:
[exclusivity] Add availability to some more test declarations.

gottesmm commented:
@swift-ci  test

gottesmm commented:
@swift-ci test apple silicon

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test apple silicon

gottesmm commented:
apple silicon test failure was something else.

gottesmm committed:
[concurrency] Disable custom executor tests on Windows since they seem to hang.

Custom executor isn't a true feature yet, so I am disabling on Windows until we
can look at why the hangs are occuring on Windows.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

mikeash commented:
For your environment variables, use `%env` to get the right prefix in all cases. Like `%env-SWIFT_DEBUG_HELP=YES`.

gottesmm commented:
I am going to let the full macOS test pass, then fix the remaining small 2 things MikeA asked for. Then I am going to do a smoke test and get this in so I can cherry-pick to 5.5.

gottesmm commented:
macOS failed due to https://github.com/apple/swift/pull/38569

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test linux platform

gottesmm commented:
Linux failure was due to that weird collection.a bug people are hitting.

gottesmm commented:
Windows was a different test failure:

  Swift(windows-x86_64) :: ClangImporter/serialization-search-paths.swift

gottesmm commented:
@swift-ci test linux platform

gottesmm commented:
@swift-ci test linux platform

gottesmm commented:
Linux just hit the collection issue. It was reverted on main so I hope I don't hit it again...

gottesmm commented:
@swift-ci clean test linux platform
